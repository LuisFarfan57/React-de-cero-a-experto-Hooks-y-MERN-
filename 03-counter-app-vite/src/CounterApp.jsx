import React, { useState } from 'react'
import PropTypes from 'prop-types'

const CounterApp = ({ value }) => {
  const [count, setCount] = useState(value)

  const handleAdd = () => {
    setCount((c) => c + 1)
  }

  const handleRest = () => {
    setCount(count - 1)
  }

  return (
    <>
      <h1 data-testid="counterAppTitle">CounterApp</h1>
      <h2> {count} </h2>
      <button onClick={handleRest}>-1</button>
      <button aria-label="btn-reset" onClick={() => setCount(value)}>Reset</button>
      <button onClick={handleAdd}>+1</button>
    </>
  )
}

CounterApp.propTypes = {
  value: PropTypes.number
}

CounterApp.defaultProps = {
  value: 0
}

export default CounterApp
