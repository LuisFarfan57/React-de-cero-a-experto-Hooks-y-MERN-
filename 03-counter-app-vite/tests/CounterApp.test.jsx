import { render, screen, fireEvent } from '@testing-library/react'
import CounterApp from '../src/CounterApp'

// describe('CounterApp test', () => {
//   it('match snapshot', () => {
//     const { container } = render(<CounterApp />)

//     expect(container).toMatchSnapshot()
//   })

//   it('should have title', () => {
//     const title = 'CounterApp'

//     const { container, getByText, getByTestId } = render(<CounterApp />)
//     expect(getByText(title)).toBeTruthy()

//     // const h1 = container.querySelector('h1')
//     // // Encuentra solo que contenga el string
//     // expect(h1.innerHTML).toContain(title)

//     expect(getByTestId('counterAppTitle').innerHTML).toContain(title)
//   })
// })

// Optimizado
describe('CounterApp test', () => {
  const title = 'CounterApp'
  const initialValue = 10

  it('match snapshot', () => {
    const { container } = render(<CounterApp />)

    expect(container).toMatchSnapshot()
  })

  it('should have title', () => {
    render(<CounterApp />)
    expect(screen.getByText(title)).toBeTruthy()
  })

  it('should have title in h1', () => {
    render(<CounterApp />)
    expect(screen.getByRole('heading', { level: 1 }).innerHTML).toContain(title)
  })

  it('should have the initial value passed by props in h2', () => {
    render(<CounterApp value={initialValue} />)
    expect(screen.getByRole('heading', { level: 2 }).innerHTML).toContain('10')
  })

  it('should increment with +1 button', () => {
    render(<CounterApp value={initialValue} />)
    fireEvent.click(screen.getByText('+1'))

    expect(screen.getByRole('heading', { level: 2 }).innerHTML).toContain('11')
  })

  it('should decrement with -1 button', () => {
    render(<CounterApp value={initialValue} />)
    fireEvent.click(screen.getByText('-1'))

    expect(screen.getByRole('heading', { level: 2 }).innerHTML).toContain('9')
  })

  it('should work the reset button', () => {
    render(<CounterApp value={initialValue} />)
    fireEvent.click(screen.getByText('+1'))
    fireEvent.click(screen.getByText('+1'))
    fireEvent.click(screen.getByText('+1'))
    fireEvent.click(screen.getByText('+1'))
    // fireEvent.click(screen.getByText('Reset'))
    fireEvent.click(screen.getByRole('button', { name: 'btn-reset' })) // Del aria-label

    expect(screen.getByRole('heading', { level: 2 }).innerHTML).toContain(initialValue.toString())
  })
})
