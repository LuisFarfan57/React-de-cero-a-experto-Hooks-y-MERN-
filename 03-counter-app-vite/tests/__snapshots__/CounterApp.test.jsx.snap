// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterApp test match snapshot 1`] = `
<div>
  <h1
    data-testid="counterAppTitle"
  >
    CounterApp
  </h1>
  <h2>
     
    0
     
  </h2>
  <button>
    -1
  </button>
  <button
    aria-label="btn-reset"
  >
    Reset
  </button>
  <button>
    +1
  </button>
</div>
`;
